### ISSUE TITLE
multi-user MyPaint!

### ISSUE THREAD

**Issue by jeisner** [mypaint/mypaint#480]
Following up on the discussion of drawing with multiple devices at once, on #259 and #479, I thought I'd fantasize about a future feature:

It would be **just awesome** to have a multi-user version of MyPaint where two or more users can collaborate remotely on the same drawing (identified by a URL).   Useful for 
* art lessons
* collaborative art or design 
* shared whiteboard for technical meetings at a distance
* drawing-based games that one could invent to be played on this platform (tic-tac-toe, variants of Pictionary, adversarial games where the drawers are trying to subvert each other's drawing ...)
* ...

A crude existing system for collaborative drawing is "A Web Whiteboard," http://awwapp.com.  MyPaint would be much better.

Each user would be running a separate MyPaint process on her own machine.  However, her process would receive messages from the other processes, reporting on other users' strokes.  It would use those messages to modify her copy of the drawing correspondingly, so that all copies are the same.  These messages would also report on other actions that modified the drawing, like "add layer" and "reposition layer."  

The messages should be relatively short because they send inking points, not pixels.  Changes to brush settings also need to be sent, but it's rare that you'd have to send a whole new brush.

Each user would have her own view (viewport + visible layers + layer opacities), her own devices, her own cursors for those devices, her own toolbar buttons, and her own keyboard.  Her toolbar buttons and keys would describe and control only her devices and her view.  She would also have her own private saved brush settings as well ... but she could use W to grab someone else's settings from the canvas.  

If she clicks on another user's icon, her copy of MyPaint should switch to that user's view of the drawing, and perhaps even show that user's cursors.  This lets her watch the other user (who may have just asked her opinion over an audio channel).  If starts drawing, then her view changes to match the one she was looking at ... otherwise, she can click on her own icon to restore her old view.  

Synchronization is the hardest thing to get right here.  All processes must interpret the strokes as if they had happened in the same order.  Perhaps I draw stroke B, and it is rendered instantly for responsiveness, but then my process gets a message that says that you drew stroke A and that A is to precede B in the global order.  In that case, my copy of MyPaint should undo B, render A, and then re-execute B for me.  But this is only necessary if A and B modify each other's pixels.  If they are in different parts of the canvas, then they commute, and my copy of MyPaint can simply render B.

(A simpler approach to synchronization is to do layer-wise locking.  If you're editing one layer, I can only edit other layers.  Only after you leave your layer am I allowed to visit it and erase or smudge your work, or perhaps undo your actions on that layer.  But this is likely to make collaboration less fun.  There are other approaches that negotiate region locking, so I'm not able to draw near your cursor.)

**Comment by jeisner** [comment]
It's not totally clear how to handle undo.  One design is that each user gets her own undo list, so that she can only undo her own actions and not other users' actions.  This means that she can undo her last stroke A even if there have been other strokes B,C,D by other users since then.  If the strokes don't overlap, then this facilitates users working in parallel on different parts of the canvas.  But if the strokes do overlap, then undoing A must be implemented (in the worst case) by undoing A,B,C,D and redoing B,C,D.  

However, this doesn't make that much sense if B smears A, for example.  So a variant design is that the undo lists aren't disjoint.  Suppose user x draws strokes A, B, and then user y draws strokes C, D, E where C overlaps A and E overlaps C.  Because of this overlap, C is inserted into x's undo list immediately following A, and E is inserted into x's undo list immediately following C.  So x has undo list [...,A,C,E,B] while y has undo list [...,C,D,E].  This means that after x undoes stroke B, she would expect to undo stroke A next, but she has been empowered to "uncover" A by undoing E and C first.  If she undoes these, they are removed from y's undo list as well, leaving y with undo list [..., D].  (It is okay that C was removed from y's undo list from underneath D, because by hypothesis, they did not overlap!)

**Comment by achadwick** [comment]
Duplicate of #67, I think.
